File: app/matching.py
"""Answer checking: typo-tolerant, accent-insensitive comparison.

Rules agreed with the user:
  - Spanish accents (ñ, á, é, í, ó, ú, ¿, ¡) are accepted both with and
    without the accent marks.
  - Small typos are tolerated; the app shows an accuracy percentage rather
    than a flat right/wrong.
"""
from __future__ import annotations

import re
import unicodedata
from dataclasses import dataclass
from difflib import SequenceMatcher

# A typed answer scoring at or above this similarity counts as "correct".
CORRECT_THRESHOLD = 0.85


def strip_accents(text: str) -> str:
    normalized = unicodedata.normalize("NFKD", text)
    return "".join(ch for ch in normalized if not unicodedata.combining(ch))


def normalize(text: str) -> str:
    text = text.strip().lower()
    text = strip_accents(text)
    text = re.sub(r"[¿?¡!.,;:]", "", text)
    text = re.sub(r"\s+", " ", text)
    return text.strip()


def similarity(a: str, b: str) -> float:
    return SequenceMatcher(None, a, b).ratio()


@dataclass
class MatchResult:
    correct: bool
    accuracy_pct: int
    best_expected: str  # which accepted answer scored highest, for feedback


def check_answer(user_answer: str, expected: str) -> MatchResult:
    """Compare a typed answer against one expected answer (accent-insensitive,
    typo-tolerant). `expected` may contain multiple accepted variants
    separated by ';' or ','.
    """
    variants = [v.strip() for v in re.split(r"[;,]", expected) if v.strip()]
    if not variants:
        variants = [expected]

    user_norm = normalize(user_answer)
    best_ratio = 0.0
    best_variant = variants[0]
    for variant in variants:
        ratio = similarity(user_norm, normalize(variant))
        if ratio > best_ratio:
            best_ratio = ratio
            best_variant = variant

    return MatchResult(
        correct=best_ratio >= CORRECT_THRESHOLD,
        accuracy_pct=round(best_ratio * 100),
        best_expected=best_variant,
    )

File: WEB/js/engine/matching.js
export const CORRECT_THRESHOLD = 0.85;

export function stripAccents(text) {
  return text.normalize("NFKD").replace(/[̀-ͯ]/g, "");
}

export function normalize(text) {
  let t = text.trim().toLowerCase();
  t = stripAccents(t);
  t = t.replace(/[¿?¡!.,;:]/g, "");
  t = t.replace(/\s+/g, " ");
  return t.trim();
}

// Longest common contiguous block between a[alo:ahi] and b[blo:bhi].
function longestMatch(a, alo, ahi, b, blo, bhi) {
  let besti = alo, bestj = blo, bestsize = 0;
  let j2len = {};
  for (let i = alo; i < ahi; i++) {
    const newj2len = {};
    for (let j = blo; j < bhi; j++) {
      if (a[i] === b[j]) {
        const k = (j2len[j - 1] || 0) + 1;
        newj2len[j] = k;
        if (k > bestsize) {
          besti = i - k + 1;
          bestj = j - k + 1;
          bestsize = k;
        }
      }
    }
    j2len = newj2len;
  }
  return [besti, bestj, bestsize];
}

function matchedLength(a, alo, ahi, b, blo, bhi) {
  const [i, j, k] = longestMatch(a, alo, ahi, b, blo, bhi);
  if (k === 0) return 0;
  let total = k;
  if (alo < i && blo < j) total += matchedLength(a, alo, i, b, blo, j);
  if (i + k < ahi && j + k < bhi) total += matchedLength(a, i + k, ahi, b, j + k, bhi);
  return total;
}

export function similarity(a, b) {
  if (a.length === 0 && b.length === 0) return 1.0;
  const m = matchedLength(a, 0, a.length, b, 0, b.length);
  return (2.0 * m) / (a.length + b.length);
}

// expected может содержать несколько принятых вариантов через ';' или ','.
export function checkAnswer(userAnswer, expected) {
  const variants = expected.split(/[;,]/).map((v) => v.trim()).filter(Boolean);
  if (variants.length === 0) variants.push(expected);

  const userNorm = normalize(userAnswer);
  let bestRatio = 0;
  let bestVariant = variants[0];
  for (const variant of variants) {
    const r = similarity(userNorm, normalize(variant));
    if (r > bestRatio) {
      bestRatio = r;
      bestVariant = variant;
    }
  }

  return {
    correct: bestRatio >= CORRECT_THRESHOLD,
    accuracyPct: Math.round(bestRatio * 100),
    bestExpected: bestVariant,
  };
}